### ISSUE TITLE
Plack::Middleware::Static or Plack::App::File doesn't serve all files

### ISSUE THREAD

**Issue by MaxPerl** [plack/Plack#586]
Hello,

I have a curious problem. I have a PSGI based, CGI::Application homepage with very much mp3 files on one page. Everything works fine, when I deploy the homepage with a dispatch CGI script. 

But if I use plackup or another PSGI server, the site serves only a few mp3 files.

Enclosed you find my PSGI file:

```
#! /usr/bin/perl -wT

# CONFIG SECTION -----------------------------------------
# --------------------------------------------------------

	#my $DOCUMENT_DIR = '../htdocs/pages/';
	#my $DATADIR = '../00_private';
	#my $DOCUMENT_ROOT = '/';
	#my $DOCUMENT_DIR = "$FindBin::Bin/../pages/";
	my $DOCUMENT_DIR = "../pages/";
	my $DATADIR = '../00_private';
	my $DOCUMENT_ROOT = '/';

# No changes should be needed below this line -------------
# ---------------------------------------------------------


use lib ("./lib/lib/perl5","./extlib5/lib/perl5", "./lib", ".");
use WebApp;
use FindBin;
use CGI::PSGI;
use Plack::App::File;
use Plack::Builder;
use Plack::Middleware::Static;

my $app = sub {
	my ($env) = @_;
	my $app = WebApp->new(QUERY => CGI::PSGI->new($env),
			TMPL_PATH => "$DATADIR/templates",
			ENV => $env,
			PARAMS => {
					'document_dir' => "$FindBin::Bin/$DOCUMENT_DIR",
					'datadir' => "$FindBin::Bin/$DATADIR",
					'document_root' => "$FindBin::Bin/$DOCUMENT_ROOT",
					'tmpl_path' => "$FindBin::Bin/$DATADIR/templates",
					'session' => $session
				},);
	
	$app->run_as_psgi();
};				
				
				
builder {
	mount "/static" => Plack::App::File->new('root' => "$FindBin::Bin/../")->to_app;
	mount "/" => builder {
		$app;
		};
};

```
Is there perhaps a maximal file size per request?
Thanks in advance,
Max

**Comment by MaxPerl** [comment]
the problem occurs in Dancer2, too...

**Comment by miyagawa** [comment]
"the site serves only a few mp3 files." is not how a bug report should be written. Describe what you expect and what you see instead. What happens to other files, are they not listed, or listed but cannot be opened, etc.?

There's no file size limit in Plack.

**Comment by MaxPerl** [comment]
Dear Tatsuhiko,
Thank you for your answer and for your help. It is difficult to describe the problem.

I have a page with about 30 elements of audio content (mp3). They are included to the page with the HTML <audio> tags. If I click to the page, I expected that every files are loaded (that means that I get 30 GET requests and the files are loaded so that you can play the files).
Instead of that only the 6-7 are loaded. The others are sometimes loaded if you click on play of the HTML player. But sometimes you cannot play these files. This happens only if I use plackup or staman. If I deploy my app with a CGI dispatch script it works wonderful! Under FCGI I have still to test. I think it works but it is curiously slower than the CGI variant.

You can see whether a file is loaded with the firefox development tools (network analysis).

I will send you a link by eMail. Then you can see the problem.

Thanks for your help,
Max

**Comment by miyagawa** [comment]
No, please do not send emails - i'm not a technical support.

You should debug on your own with browser consoles, or using non-browser tools like curl or telnet. I would check file permissions, or some additional file headers or buffer size if there's any.

**Comment by MaxPerl** [comment]
oh sorry I just sended. I didn't want to publish the link here public. Sorry. I didn't think that you are something like a technical support. I just thought, so could it be easier for you to see the problem...

Thanks for the tips. I will test them...

**Comment by MaxPerl** [comment]
PS.: If I just open one file. Every file works as expected..

**Comment by miyagawa** [comment]
so you're having a simultaneous connection problem, rather than being unable to open some files, not all. plackup runs as a single process server (HTTP::Server::PSGI) by default, so if your browser opens one connection to one file, then you cannot open another connection to it.

You're recommended to run multi-process servers such as Starman and maybe front it with a reverse proxy server such as nginx, on production.

**Comment by MaxPerl** [comment]
Thank you very much for your great help. To be honest, the details are difficult to understand for me. I am just a hobby programmer.. Perhaps the plack interface is for more professional people..

I tried Starman, too (starman -l 127.0.0.1:5000 --workers 10 ./app.psgi). But without effort.

The reverse proxy solution I don't understand at the moment, but I will read in it... I don't think that I can go this way, because I am at a shared hoster, have no sudo rights and can only edit .htaccess (that means no VirtualHosts, ProxyPass etc. options).
At the moment my .htaccess is very simple:
```
RewriteEngine On
RewriteRule ^(.*) http://127.0.0.1:5000/$1 [P,L]

```

I noticed that I have the same problem with FastCGI.
At the moment I solved it mostly by adding preload="none" to the audio tags. Largely playing now works (but sometimes it still does not work). At worst I must use the good old CGI interface.

After all I don't think that this is a plack issue but more a setting problem, so I close this issue.

Thanks for all,
Max
